Merge 3.2.x into 4.0.x

greg0ire opened the pull request.

morozov committed:
Rework connect/disconnect routines in functional tests

$ git grep --name-only 'extends FunctionalTestCase' | xargs sed -i -E 's/parent::(setUp|tearDown)\\(\\);//g'

morozov committed:
Remove some shared state from functional tests

derrabus committed:
Merge pull request #4868 from morozov/functional-connect-disconnect

 Rework connect/disconnect routines in functional tests

morozov committed:
Merge branch '3.1.x' into 3.2.x

kitloong committed:
Get view definition for SQL Server

Currently, SQLServer2012Platform::getListViewsSQL does not export view definition.
This commit get view definition for SQL server.
Reference: https://docs.microsoft.com/en-us/sql/relational-databases/views/get-information-about-a-view?view=sql-server-ver15#to-get-the-definition-and-properties-of-a-view

morozov committed:
Merge pull request #4844 from kitloong/fix-sqlserver-view

Get view definition for SQL Server

morozov committed:
Bump PHPStan level to 8

greg0ire committed:
Get rid of current working directory variable

They are useless because there is only one directory from which this
configuration makes sense.

derrabus committed:
Merge pull request #4873 from morozov/phpstan-lvl-8

Bump PHPStan level to 8

derrabus committed:
Run tests on PHP 8.1 with different database drivers

derrabus committed:
Catch MySQLi exceptions on PHP 8.1

derrabus committed:
Fix XSD paths in PHPUnit XML config files

greg0ire committed:
Merge pull request #4874 from greg0ire/get-rid-of-cwd

Get rid of current working directory variable

greg0ire committed:
Merge remote-tracking branch 'origin/2.13.x' into 3.1.x

morozov committed:
Merge pull request #4875 from derrabus/ci/database-drivers-php-8.1

Run tests on PHP 8.1 with different database drivers

greg0ire committed:
Merge remote-tracking branch 'origin/3.1.x' into 3.2.x

greg0ire committed:
Merge remote-tracking branch 'origin/2.13.x' into 3.1.x

greg0ire committed:
Add missing upcast

derrabus committed:
Merge pull request #4876 from derrabus/bugfix/xsd-paths

Fix XSD paths in PHPUnit XML config files

greg0ire committed:
Merge pull request #4880 from derrabus/bugfix/mysql-exceptions

Catch MySQLi exceptions on PHP 8.1

greg0ire committed:
Merge remote-tracking branch 'origin/3.1.x' into 3.1.x

greg0ire committed:
Reuse centralized coding standard workflow

Other workflows have divergences that make this impossible:
- continuous integration has special jobs because we support many RDBMS;
- static analysis has an extra command to test files under a special
  directory;
- release has a lot fewer steps than its centralized counterpart.

greg0ire committed:
Make $flags argument non-nullable

Since php 8.1, passing a null $flags argument to mysqli::real_connect()
is deprecated.

greg0ire committed:
Suppress errors with non-obvious fixes

greg0ire committed:
Merge pull request #4883 from greg0ire/reuse-workflows

Reuse centralized coding standard workflow

greg0ire committed:
Merge remote-tracking branch 'origin/2.13.x' into 3.1.x

greg0ire committed:
Merge remote-tracking branch 'origin/3.1.x' into 3.2.x

greg0ire commented:
@morozov any hint about the MySQL failures?

morozov commented:
I fixed the `pdo_mysql` issue, see the details in the update. The issue wasn't reproducible on `mysqli` because apparently, it returns true when rolling back a non-existing transaction:
```php
<?php

var_dump(mysqli_connect('127.0.0.1', 'root', '', 'mysql')->rollback());
// bool(true)
```
UPD: reported ^^^ as https://bugs.php.net/bug.php?id=81533.


The remaining issue is caused by the return value change in the`PDO::lastInsertId()` implementation of `pdo_sqlsrv` (https://github.com/microsoft/msphpsql/pull/1282). Now it returns `false` when there's no last id:
> Returns the last insert id as a string. (if PHP_VERSION_ID < 80100)

We need to change `PDOConnection::lastInsertId()`:
```diff
-        if ($value === '0' || $value === '') {
+        if ($value === '0' || $value === '' || $value === false) {
```

Although `PDO::lastInsertId()` has been returning `string|false` since forever, it looks like it wasn't reflected in the PHP documentation until https://github.com/php/doc-en/pull/494. So, we need to suppress all false-positives:

Psalm:
```
ERROR: DocblockTypeContradiction - src/Driver/PDO/Connection.php:118:48 - string does not contain false (see https://psalm.dev/155)
        if ($value === '0' || $value === '' || $value === false) {
```
PHPStan:
```
 ------ -----------------------------------------------------------------------------------------------
  Line   src/Driver/PDO/Connection.php
 ------ -----------------------------------------------------------------------------------------------
  118    Strict comparison using === between non-empty-string and false will always evaluate to false.
 ------ -----------------------------------------------------------------------------------------------
```
And contribute the update to all the stubs we use (Jetbrains, ~~PHPStan~~ (not stubbed), Psalm). What a mess…

derrabus committed:
Remove mysqli_report() call

Signed-off-by: Alexander M. Turek <me@derrabus.de>

greg0ire commented:
Thanks for investigating this, I can add a fix to this PR I suppose. Any idea why it affects only the 4.0.x branch?

> Although PDO::lastInsertId() has been returning string|false since forever

:thinking: I thought it was since 8.1

For Jetbrains, it looks OK already : https://github.com/JetBrains/phpstorm-stubs/pull/1234
For PHPStan, the issue goes away when I disable strict rules

greg0ire commented:
I have tweaked the settings and applied your suggestion @morozov

greg0ire commented:
Replaced a link to the issue with a link to the fix (https://github.com/phpstan/phpstan-src/pull/719)

morozov commented on phpstan.neon.dist:
```diff
@@ -161,7 +161,48 @@ parameters:
-                - %currentWorkingDirectory%/src/DriverManager.php
+                - src/DriverManager.php
+
+        # See https://github.com/doctrine/dbal/pull/4707
```
Looks like a conflict resolution mistake. See https://github.com/doctrine/dbal/pull/4722.

morozov commented on phpstan.neon.dist:
```diff
@@ -161,7 +161,48 @@ parameters:
+                - src/Schema/Comparator.php
+                - src/Schema/Schema.php
+
+        # We're checking for invalid invalid input
```
This one as well. See https://github.com/doctrine/dbal/pull/4626.

morozov commented on phpstan.neon.dist:
```diff
@@ -161,7 +161,48 @@ parameters:
+            count: 1
+            path: src/Cache/QueryCacheProfile.php
+        -
+            message: '~^Method Doctrine\\DBAL\\Platforms\\AbstractPlatform::escapeStringForLike\(\) should return string but returns string\|null\.$~'
```
The reference to the PHPStan issue is missing from the original commit. See https://github.com/doctrine/dbal/pull/4873.

morozov commented on phpstan.neon.dist:
```diff
@@ -161,7 +161,48 @@ parameters:
+
+        # Requires a release of https://github.com/phpstan/phpstan-src/pull/719
+        -
+            message: '~Strict comparison using \=\=\= between non\-empty\-string and false will always evaluate to false\.~'
```
Why do `=` and `-` need escaping?

greg0ire replied:
I don't know, I used `preg_quote`

morozov replied:
I believe `preg_quote()` is unnecessarily precautious. These characters have special meaning in certain contexts but not by default.

morozov commented on psalm.xml.dist:
```diff
@@ -74,6 +74,10 @@
                 <file name="src/Connection.php"/>
                 <file name="src/Driver/IBMDB2/Statement.php"/>
+                <!--
+                    Requires a release of https://github.com/JetBrains/phpstorm-stubs/pull/1234
```
The phpDoc for `PDO::lastInsertId()` there still contains:
```
* @return string If a sequence name was not specified for the <i>name</i>
```
The return type in the method signature may be insufficient since it's preceded by a one-line annotation.

greg0ire replied:
:thinking: right… what should we do?

morozov replied:
Just send another patch to `JetBrains/phpstorm-stubs` and replace the link.

morozov commented on phpstan.neon.dist:
```diff
@@ -161,7 +161,35 @@ parameters:
-                - %currentWorkingDirectory%/src/DriverManager.php
+                - src/DriverManager.php
+
+	# https://github.com/phpstan/phpstan/issues/1901
```
This is indented incorrectly.

morozov commented:
> Any idea why it affects only the 4.0.x branch?

My guess is that rollback failures on PDO prior to 4.0.x don't result in an exception but I haven't checked.

> I thought it was since 8.1

The code that returns false in the case if there's no last insert id exists since https://github.com/php/php-src/commit/2c5b2fc105bd6af6571c2bd39d2785c627c6e578. But it looks like none of the PDO-based extensions respect the API (except for `pdo_sqlsrv` since recently).

morozov commented on phpstan.neon.dist:
```diff
@@ -1,8 +1,8 @@
 parameters:
     level: 7
```
And now we're back to level 7.

greg0ire replied:
:facepalm: sorry, I don't know where this went so wrong… I will give this another try later.

derrabus committed:
Merge pull request #4887 from derrabus/remove/mysqli-report

greg0ire committed:
Bump PHPStan level to 8

It has been wrongly reset to 7 during a faulty merge up.

greg0ire committed:
Merge pull request #4891 from greg0ire/phpstan-lvl-8

Bump PHPStan level to 8

greg0ire committed:
Add null-checks where needed

morozov commented on tests/Functional/Schema/MySQL/ComparatorTest.php:
```diff
@@ -15,6 +15,8 @@
 use Doctrine\DBAL\Tests\FunctionalTestCase;
 use Doctrine\DBAL\Types\Types;
 
+use function assert;
```
We don't normally use PHP assertions in PHPUnit tests.

morozov commented on tests/Functional/Schema/MySQL/ComparatorTest.php:
```diff
@@ -95,7 +93,9 @@ private function createLobTable(string $type, int $length): Table
         $column = $table->getColumn('lob');
-        $column->setLength($column->getLength() + 1);
+        $length = $column->getLength();
+        self::assertNotNull($length);
```
Would it make sense to accept `int $length` in this method and rename it to say `setBlobLength()`? It feels awkward that the test makes assertions against its own implementation.

morozov commented on tests/Functional/Schema/SQLServerSchemaManagerTest.php:
```diff
@@ -203,6 +203,7 @@ public function testPkOrdering(): void
 
         self::assertCount(1, $indexes);
         $firstIndex = array_shift($indexes);
+        self::assertNotNull($firstIndex);
```
This could be logged as a bug/improvement to PHPStan. `array_shift()` called on a non-empty `Index[]` (which is asserted by the `assertCount()` above) cannot return null.

greg0ire replied:
https://github.com/phpstan/phpstan/issues/5803

greg0ire replied:
It got closed. I also discovered about `\PHPStan\dumpType()`, and apparently, the actual issue is not with `array_shift`, but with `assertCount`, which does not transmit the `&nonEmpty` part. Using a normal `assert` does.

greg0ire replied:
`assertNotEmpty` does not help either: https://github.com/phpstan/phpstan-phpunit/issues/54

greg0ire replied:
I added a comment but I think it's not worth creating a separate issue for this.

morozov replied:
Thanks! It's fine as long as we know the root cause and the issue is documented upstream.

morozov reviewed (commented):
There's just one cosmetic nitpick. Otherwise, looks good.

I'd like to avoid a failing build on the merge commit. Could you cherry-pick the last commit into `4.0.x` before the merge and then redo the merge? Going through all the conflict resolution could be avoided if before redoing the merge the commit of the desired tree is backed up and then checked out.

greg0ire committed:
Refactor into setter to avoid assertion

Because getLength() may return null, just setting the length to its
result + 1 is unsafe, and it is more simple to set the length
directly.

morozov reviewed (approved):
I believe the commit that reworks the `increaseLobLength()` method could go to `4.0.x` prior to the merge as well. Otherwise, there are multiple logically conflicting changes to the same method in the same PR. But I don't think it's a blocker. Also, my fault, the new method should be named as `setLobLength()` not `setBlobLength()`.

greg0ire committed:
Merge remote-tracking branch 'origin/3.2.x' into 4.0.x
